frontend/src/StealthHome.tsx

```diff
@@ -0,0 +1,157 @@
+import Button from "@material-ui/core/Button";
+import * as firebase from "firebase/app";
+import "firebase/auth";
+import { WelcomePage } from "./WelcomePage";
+import { yellow } from "@material-ui/core/colors";
+
+const styles = makeStyles((theme: Theme) =>
+  createStyles({
+    root: {
+      flexGrow: 1,
+    },
+    menuButton: {
+      marginRight: theme.spacing(2),
+    },
+    actionButton: {
+      marginLeft: "auto",
+      color: theme.palette.getContrastText(yellow[700]),
+      backgroundColor: yellow[700],
+      "&:hover": {
+        backgroundColor: yellow[900],
+      },
+    },
+    logo: {
+      height: "47px",
+    },
+    topBar: {
+      backgroundColor: "#2b5482",
+    },
+    bottomBar: {
+      backgroundColor: "#2b5482",
+      top: "auto",
+      bottom: 0,
+      height: "47px",
+    },
+  })
+);
+
+export type StealthHomeProps = {
+  children: React.ReactNode;
+};
+
+const defaultContext = {
+  user: (undefined as unknown) as firebase.User,
+  setUserTopButton: (button: React.ReactNode | null) => {},
+};
+export const StealthContext = React.createContext(defaultContext);
+
+type StealthTopButtonProps = {
+  children: React.ReactNode;
+  onClick: (e: unknown) => void;
+};
+export const StealthTopButton: React.FC<StealthTopButtonProps> = ({
+  children,
+  onClick,
+}) => {
+  const classes = styles();
+
+  return (
+    <Button
+      variant="contained"
+      color="inherit"
+      id="loginState"
+      className={classes.actionButton}
+      onClick={onClick}
+    >
+      {children}
+    </Button>
+  );
+};
+
+export const StealthHome: React.FC<StealthHomeProps> = ({ children }) => {
+  const classes = styles();
+  // user has three possible settings: the user, null (signed out), undefined (dunno)
+  // the page is rendered differently in each case
+  const [user, setUser] = useState<firebase.User | null | undefined>(undefined);
+
+  const [topButton, setTopButton] = useState<React.ReactNode | null>(null);
+  const [userTopButton, setUserTopButton] = useState<React.ReactNode | null>(
+    null
+  );
+
+  useEffect(() => {
+    (async () => {
+      const result = await firebase.auth().getRedirectResult();
+      if (result.user) {
+        // User just signed in. Can get result.credential and result.credential.accessToken
+        console.log("Case 1 result.user");
+        setUser(result.user);
+      } else if (firebase.auth().currentUser) {
+        // User already signed in
+        console.log("Case 2 result.user");
+        setUser(firebase.auth().currentUser);
+      } else {
+        setUser(null);
+      }
+    })();
+  }, []);
+  useEffect(() => {
+    const toggleSignIn = () => {
+      if (!user) {
+        console.log("sign in");
+        const provider = new firebase.auth.GoogleAuthProvider();
+        firebase.auth().signInWithRedirect(provider);
+      } else {
+        firebase.auth().signOut();
+        setUser(null);
+      }
+    };
+
+    setTopButton(
+      userTopButton ||
```

Comment:
this nested logic is hard to follow, consider breaking into several statements.

Reply:
Removed.